### PR TITLE
Widget

### PR DESCRIPTION
**Pull request by kson1128**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+const showDisplayZoom = viewportZoomLevel => {
+  const displayZoom = (viewportZoomLevel * ZOOM_MULTIPLIER).toFixed(2);
+  return `Zoom Level: ${displayZoom}x`;
+};
+
+//Extract zoom value from
+const getViewportZoomLevel = zoomLvl => {
+  return parseInt(zoomLvl) / ZOOM_MULTIPLIER;
+};
```
Comment (chunloy) [inline index.js:16]: move helper functions below constants

File: index.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+customsButtonsContainer.id = 'customsButtonsContainer';
+
+const viewer = OpenSeadragon({
+  id: 'osd',
+  prefixUrl: 'https://learn.mghpathology.org/resources/assets/osd_icons/',
+  tileSources:
+    'https://learn.mghpathology.org/pv-http/WSI22-98-Case-5_1.mrxs.dzi',
+
+  minZoomPixelRatio: 2,
+});
```
Comment (chunloy) [inline index.js:29]: Magic strings + numbers, factor out to constant. Might be easier to read if you factor out the object as well.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+const createZoomButtons = () => {
+  ZOOM_LEVEL_TEXT.forEach(zoomLevel => {
+    const newButton = document.createElement('button');
+    newButton.className = 'preset_button';
+    newButton.textContent = zoomLevel;
+
+    newButton.addEventListener('click', event => {
+      viewer.viewport.zoomTo(getViewportZoomLevel(event.target.innerHTML));
+    });
```
Comment (chunloy) [inline index.js:45]: move helper function up

File: index.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  minZoomPixelRatio: 2,
+});
+
+const _zoomLevel = document.createElement('div');
+_zoomLevel.id = 'zoomLevel';
+
+viewer.container.append(_zoomLevel);
+viewer.container.style['border'] = '1px solid black';
+
+const createZoomButtons = () => {
```
Comment (chunloy) [inline index.js:37]: move helper function up

File: index.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    const newButton = document.createElement('button');
+    newButton.className = 'preset_button';
+    newButton.textContent = zoomLevel;
+
+    newButton.addEventListener('click', event => {
+      viewer.viewport.zoomTo(getViewportZoomLevel(event.target.innerHTML));
+    });
+
+    viewer.addButton(
+      new OpenSeadragon.Button({
```
Comment (chunloy) [inline index.js:48]: Nit: object is small, but could factor out for readability.

File: index.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+const viewer = OpenSeadragon({
+  id: 'osd',
+  prefixUrl: 'https://learn.mghpathology.org/resources/assets/osd_icons/',
+  tileSources:
+    'https://learn.mghpathology.org/pv-http/WSI22-98-Case-5_1.mrxs.dzi',
+
+  minZoomPixelRatio: 2,
+});
+
+const _zoomLevel = document.createElement('div');
```
Comment (chunloy) [inline index.js:31]: Not sure I understand the reasoning behind prefixing variable names with underscores. Typically, the underscore prefix represents a private method or member (OOP).

File: README.md
Comment (chunloy) [inline README.md:1]: Be sure to add screenshots, or link the demo here.

File: index.js
Comment (chunloy) [inline index.js:1]: order of operations is still a bit confusing, needs cleanup. I'd suggest following this order:

1. declare constants
2. helper functions
3. create elements
4. add handlers/listeners

File: index.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+_zoomLevel.id = 'zoomLevel';
+
+viewer.container.append(_zoomLevel);
+viewer.container.style['border'] = '1px solid black';
+
+const createZoomButtons = () => {
+  ZOOM_LEVEL_TEXT.forEach(zoomLevel => {
+    const newButton = document.createElement('button');
```
Comment (chunloy) [inline index.js:39]: rename variable, maybe zoomButton?

File: server.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+const express = require('express');
+const path = require('path');
+const app = express();
+const cors = require('cors');
+const port = 8000;
+
+app.use(
+  cors({
+    origin: 'http://localhost:8000',
```
Comment (chunloy) [inline server.js:9]: use port variable

File: server.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  })
+);
+
+app.use(
```
Comment (chunloy) [inline server.js:13]: would be helpful to see a comment describing what this is for

File: style.css
Comment (chunloy) [inline style.css:1]: Aside from `#osd`, is it necessary to use ids?

File: style.css [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+#customsButtonsContainer {
+  position: absolute;
+  z-index: 10;
+  left: 10px;
+  top: 50px;
+}
+
+#presets {
```
Comment (chunloy) [inline style.css:20]: I don't think this is being used?

File: style.css [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  padding: 8px;
+  z-index: 20;
+  width: fit-content;
+  right: 0px;
+  margin: 6px;
+  background-color: #ffffff9e;
+  border-radius: 8px;
+}
+
+.show_presets {
```
Comment (chunloy) [inline style.css:59]: is this being used?

File: style.css [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  top: 50px;
+}
+
+#presets {
+  position: relative;
+  display: flex;
+  justify-content: center;
+}
+
+.preset_button {
```
Comment (chunloy) [inline style.css:26]: needs a better name, maybe zoom_button?

File: index.js (hunk 2 above)
Comment (kson1128) [reply index.js:29]: fixed

**Review by chunloy — COMMENTED** [review]
Good start, left some comments.

File: index.js (hunk 6 above)
Comment (kson1128) [reply index.js:31]: fixed to zoomLevelDisplay

File: index.js (hunk 1 above)
Comment (kson1128) [reply index.js:16]: fixed

File: index.js (hunk 5 above)
Comment (kson1128) [reply index.js:48]: fixed

File: index.js (hunk 7 above)
Comment (kson1128) [reply index.js:39]: fixed

File: server.js (hunk 8 above)
Comment (kson1128) [reply server.js:9]: fixed

File: index.js (hunk 4 above)
Comment (kson1128) [reply index.js:37]: fixed

File: style.css (hunk 10 above)
Comment (kson1128) [reply style.css:20]: removed

File: style.css (hunk 11 above)
Comment (kson1128) [reply style.css:59]: fixed

File: style.css [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+#osd {
+  position: relative;
+  width: min(100%, 50rem);
+  height: 700px;
+  margin: 20px auto 0;
+}
+
+#customsButtonsContainer {
```
[the remark this answers is not in the document]
Comment (kson1128) [reply style.css:13]: removed

**Pull request merged**